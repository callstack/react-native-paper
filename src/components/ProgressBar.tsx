import * as React from 'react';
import {
  Animated,
  I18nManager,
  LayoutChangeEvent,
  Platform,
  StyleProp,
  StyleSheet,
  View,
  ViewStyle,
} from 'react-native';

import setColor from 'color';

import { useInternalTheme } from '../core/theming';
import type { ThemeProp } from '../types';

export type Props = React.ComponentPropsWithRef<typeof View> & {
  /**
   * Animated value (between 0 and 1). This tells the progress bar to rely on this value to animate it.
   * Note: It should not be used in parallel with the `progress` prop.
   */
  animatedValue?: number;
  /**
   * Progress value (between 0 and 1).
   * Note: It should not be used in parallel with the `animatedValue` prop.
   */
  progress?: number;
  /**
   * Color of the progress bar. The background color will be calculated based on this but you can change it by passing `backgroundColor` to `style` prop.
   */
  color?: string;
  /**
   * If the progress bar will show indeterminate progress.
   */
  indeterminate?: boolean;
  /**
   * Whether to show the ProgressBar (true, the default) or hide it (false).
   */
  visible?: boolean;
  /**
   * Style of filled part of the ProgresBar.
   */
  fillStyle?: Animated.WithAnimatedValue<StyleProp<ViewStyle>>;
  style?: StyleProp<ViewStyle>;
  /**
   * @optional
   */
  theme?: ThemeProp;
  /**
   * testID to be used on tests.
   */
  testID?: string;
};

const INDETERMINATE_DURATION = 2000;
const INDETERMINATE_MAX_WIDTH = 0.6;
const { isRTL } = I18nManager;

/**
 * Progress bar is an indicator used to present progress of some activity in the app.
 *
 * ## Usage
 * ```js
 * import * as React from 'react';
 * import { ProgressBar, MD3Colors } from 'react-native-paper';
 *
 * const MyComponent = () => (
 *   <ProgressBar progress={0.5} color={MD3Colors.error50} />
 * );
 *
 * export default MyComponent;
 * ```
 */
const ProgressBar = ({
  color,
  indeterminate,
  progress = 0,
  visible = true,
  theme: themeOverrides,
  animatedValue,
  style,
  fillStyle,
  testID = 'progress-bar',
  ...rest
}: Props) => {
  const isWeb = Platform.OS === 'web';
  const theme = useInternalTheme(themeOverrides);
  const { current: timer } = React.useRef<Animated.Value>(
    new Animated.Value(0)
  );
  const { current: fade } = React.useRef<Animated.Value>(new Animated.Value(0));
  const passedAnimatedValue =
    React.useRef<Props['animatedValue']>(animatedValue);
  const [width, setWidth] = React.useState<number>(0);
  const [prevWidth, setPrevWidth] = React.useState<number>(0);

  const indeterminateAnimation =
    React.useRef<Animated.CompositeAnimation | null>(null);

  const { scale } = theme.animation;

  React.useEffect(() => {
    passedAnimatedValue.current = animatedValue;
  });

  const startAnimation = React.useCallback(() => {
    // Show progress bar
    Animated.timing(fade, {
      duration: 200 * scale,
      toValue: 1,
      useNativeDriver: true,
      isInteraction: false,
    }).start();

    /**
     * We shouldn't add @param animatedValue to the
     * deps array, to avoid the unnecessary loop.
     * We can only check if the prop is passed initially,
     * and we do early return.
     */
    const externalAnimation =
      typeof passedAnimatedValue.current !== 'undefined' &&
      passedAnimatedValue.current >= 0;

    if (externalAnimation) {
      return;
    }

    // Animate progress bar
    if (indeterminate) {
      if (!indeterminateAnimation.current) {
        indeterminateAnimation.current = Animated.timing(timer, {
          duration: INDETERMINATE_DURATION,
          toValue: 1,
          // Animated.loop does not work if useNativeDriver is true on web
          useNativeDriver: !isWeb,
          isInteraction: false,
        });
      }

      // Reset timer to the beginning
      timer.setValue(0);

      Animated.loop(indeterminateAnimation.current).start();
    } else {
      Animated.timing(timer, {
        duration: 200 * scale,
        toValue: progress ? progress : 0,
        useNativeDriver: true,
        isInteraction: false,
      }).start();
    }
  }, [fade, scale, indeterminate, timer, progress, isWeb]);

  const stopAnimation = React.useCallback(() => {
    // Stop indeterminate animation
    if (indeterminateAnimation.current) {
      indeterminateAnimation.current.stop();
    }

    Animated.timing(fade, {
      duration: 200 * scale,
      toValue: 0,
      useNativeDriver: true,
      isInteraction: false,
    }).start();
  }, [fade, scale]);

  React.useEffect(() => {
    if (visible) startAnimation();
    else stopAnimation();
  }, [visible, startAnimation, stopAnimation]);

  React.useEffect(() => {
    if (animatedValue && animatedValue >= 0) {
      timer.setValue(animatedValue);
    }
  }, [animatedValue, timer]);

  React.useEffect(() => {
    // Start animation the very first time when previously the width was unclear
    if (visible && prevWidth === 0) {
      startAnimation();
    }
  }, [prevWidth, startAnimation, visible]);

  const onLayout = (event: LayoutChangeEvent) => {
    setPrevWidth(width);
    setWidth(event.nativeEvent.layout.width);
  };

  const tintColor = color || theme.colors?.primary;
  const trackTintColor = theme.isV3
    ? theme.colors.surfaceVariant
    : setColor(tintColor).alpha(0.38).rgb().string();

  return (
    <View
      onLayout={onLayout}
      {...rest}
      accessible
      accessibilityRole="progressbar"
      accessibilityState={{ busy: visible }}
      accessibilityValue={
        indeterminate
          ? {}
          : { min: 0, max: 100, now: Math.round(progress * 100) }
      }
      style={isWeb && styles.webContainer}
      testID={testID}
    >
      <Animated.View
        style={[
          styles.container,
          { backgroundColor: trackTintColor, opacity: fade },
          style,
        ]}
      >
        {width ? (
          <Animated.View
            testID={`${testID}-fill`}
            style={[
              styles.progressBar,
              {
                width,
                backgroundColor: tintColor,
                transform: [
                  {
                    translateX: timer.interpolate(
                      indeterminate
                        ? {
                            inputRange: [0, 0.5, 1],
                            outputRange: [
                              (isRTL ? 1 : -1) * 0.5 * width,
                              (isRTL ? 1 : -1) *
                                0.5 *
                                INDETERMINATE_MAX_WIDTH *
                                width,
                              (isRTL ? -1 : 1) * 0.7 * width,
                            ],
                          }
                        : {
                            inputRange: [0, 1],
                            outputRange: [(isRTL ? 1 : -1) * 0.5 * width, 0],
                          }
                    ),
                  },
                  {
                    // Workaround for workaround for https://github.com/facebook/react-native/issues/6278
                    scaleX: timer.interpolate(
                      indeterminate
                        ? {
                            inputRange: [0, 0.5, 1],
                            outputRange: [
                              0.0001,
                              INDETERMINATE_MAX_WIDTH,
                              0.0001,
                            ],
                          }
                        : {
                            inputRange: [0, 1],
                            outputRange: [0.0001, 1],
                          }
                    ),
                  },
                ],
              },
              fillStyle,
            ]}
          />
        ) : null}
      </Animated.View>
    </View>
  );
};

const styles = StyleSheet.create({
  container: {
    height: 4,
    overflow: 'hidden',
  },
  webContainer: {
    width: '100%',
    height: '100%',
  },
  progressBar: {
    flex: 1,
  },
});

export default ProgressBar;
