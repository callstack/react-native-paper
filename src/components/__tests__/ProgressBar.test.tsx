import * as React from 'react';
import { Animated, Platform, StyleSheet } from 'react-native';

import { render, waitFor } from '@testing-library/react-native';

import ProgressBar, { Props } from '../ProgressBar';

const layoutEvent = {
  nativeEvent: {
    layout: {
      width: 100,
    },
  },
};

const styles = StyleSheet.create({
  fill: {
    borderRadius: 4,
  },
});

const a11yRole = 'progressbar';

class ClassProgressBar extends React.Component<Props> {
  render() {
    return <ProgressBar {...this.props} />;
  }
}

const AnimatedProgressBar = Animated.createAnimatedComponent(ClassProgressBar);

afterEach(() => {
  Platform.OS = 'ios';
});

it('renders progress bar with animated value', async () => {
  const tree = render(<AnimatedProgressBar animatedValue={0.2} />);
  await waitFor(() => tree.getByRole(a11yRole).props.onLayout(layoutEvent));

  tree.update(<AnimatedProgressBar animatedValue={0.4} />);

  expect(tree.container.props['animatedValue']).toBe(0.4);
});

it('renders progress bar with specific progress', async () => {
  const tree = render(<ProgressBar progress={0.2} />);
  await waitFor(() => tree.getByRole(a11yRole).props.onLayout(layoutEvent));

  expect(tree.toJSON()).toMatchSnapshot();
});

it('renders hidden progress bar', async () => {
  const tree = render(<ProgressBar progress={0.2} visible={false} />);
  await waitFor(() => tree.getByRole(a11yRole).props.onLayout(layoutEvent));

  expect(tree.toJSON()).toMatchSnapshot();
});

it('renders colored progress bar', async () => {
  const tree = render(<ProgressBar progress={0.2} color="red" />);
  await waitFor(() => tree.getByRole(a11yRole).props.onLayout(layoutEvent));

  expect(tree.toJSON()).toMatchSnapshot();
});

it('renders indeterminate progress bar', async () => {
  const tree = render(<ProgressBar indeterminate />);
  await waitFor(() => tree.getByRole(a11yRole).props.onLayout(layoutEvent));

  expect(tree.toJSON()).toMatchSnapshot();
});

it('renders progress bar with full height on web', () => {
  Platform.OS = 'web';
  const tree = render(<ProgressBar progress={0.2} />);

  expect(tree.getByRole(a11yRole)).toHaveStyle({
    width: '100%',
    height: '100%',
  });
});

it('renders progress bar with custom style of filled part', async () => {
  const tree = render(
    <ProgressBar progress={0.2} fillStyle={styles.fill} testID="progress-bar" />
  );
  await waitFor(() => tree.getByRole(a11yRole).props.onLayout(layoutEvent));

  expect(tree.getByTestId('progress-bar-fill')).toHaveStyle({
    borderRadius: 4,
  });
});
