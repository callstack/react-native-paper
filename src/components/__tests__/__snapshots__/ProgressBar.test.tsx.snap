// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders colored progress bar 1`] = `
<View
  accessibilityRole="progressbar"
  accessibilityState={
    {
      "busy": true,
    }
  }
  accessibilityValue={
    {
      "max": 100,
      "min": 0,
      "now": 20,
    }
  }
  accessible={true}
  onLayout={[Function]}
  style={false}
  testID="progress-bar"
>
  <View
    collapsable={false}
    style={
      {
        "backgroundColor": "rgba(231, 224, 236, 1)",
        "height": 4,
        "opacity": 0,
        "overflow": "hidden",
      }
    }
  >
    <View
      collapsable={false}
      style={
        {
          "backgroundColor": "red",
          "flex": 1,
          "transform": [
            {
              "translateX": -50,
            },
            {
              "scaleX": 0.0001,
            },
          ],
          "width": 100,
        }
      }
      testID="progress-bar-fill"
    />
  </View>
</View>
`;

exports[`renders hidden progress bar 1`] = `
<View
  accessibilityRole="progressbar"
  accessibilityState={
    {
      "busy": false,
    }
  }
  accessibilityValue={
    {
      "max": 100,
      "min": 0,
      "now": 20,
    }
  }
  accessible={true}
  onLayout={[Function]}
  style={false}
  testID="progress-bar"
>
  <View
    collapsable={false}
    style={
      {
        "backgroundColor": "rgba(231, 224, 236, 1)",
        "height": 4,
        "opacity": 0,
        "overflow": "hidden",
      }
    }
  >
    <View
      collapsable={false}
      style={
        {
          "backgroundColor": "rgba(103, 80, 164, 1)",
          "flex": 1,
          "transform": [
            {
              "translateX": -50,
            },
            {
              "scaleX": 0.0001,
            },
          ],
          "width": 100,
        }
      }
      testID="progress-bar-fill"
    />
  </View>
</View>
`;

exports[`renders indeterminate progress bar 1`] = `
<View
  accessibilityRole="progressbar"
  accessibilityState={
    {
      "busy": true,
    }
  }
  accessibilityValue={{}}
  accessible={true}
  onLayout={[Function]}
  style={false}
  testID="progress-bar"
>
  <View
    collapsable={false}
    style={
      {
        "backgroundColor": "rgba(231, 224, 236, 1)",
        "height": 4,
        "opacity": 0,
        "overflow": "hidden",
      }
    }
  >
    <View
      collapsable={false}
      style={
        {
          "backgroundColor": "rgba(103, 80, 164, 1)",
          "flex": 1,
          "transform": [
            {
              "translateX": -50,
            },
            {
              "scaleX": 0.0001,
            },
          ],
          "width": 100,
        }
      }
      testID="progress-bar-fill"
    />
  </View>
</View>
`;

exports[`renders progress bar with specific progress 1`] = `
<View
  accessibilityRole="progressbar"
  accessibilityState={
    {
      "busy": true,
    }
  }
  accessibilityValue={
    {
      "max": 100,
      "min": 0,
      "now": 20,
    }
  }
  accessible={true}
  onLayout={[Function]}
  style={false}
  testID="progress-bar"
>
  <View
    collapsable={false}
    style={
      {
        "backgroundColor": "rgba(231, 224, 236, 1)",
        "height": 4,
        "opacity": 0,
        "overflow": "hidden",
      }
    }
  >
    <View
      collapsable={false}
      style={
        {
          "backgroundColor": "rgba(103, 80, 164, 1)",
          "flex": 1,
          "transform": [
            {
              "translateX": -50,
            },
            {
              "scaleX": 0.0001,
            },
          ],
          "width": 100,
        }
      }
      testID="progress-bar-fill"
    />
  </View>
</View>
`;
